Add actions for Hockeypuck

swetha1654 opened the pull request:
Applicable spec: [Hockeypuck K8s Charm Spec](https://docs.google.com/document/d/1cGJWMD86Q5kwx1AtBkJ4D5eRcb31w2SE2js3IPVRF2U/edit?tab=t.0)

### Overview

<!-- A high level overview of the change -->
1. Added the following 2 actions for Hockeypuck:
    - blacklist-and-delete-key
    - rebuild-prefix-tree

2. Added integration tests for the same

### Rationale

<!-- The reason the change is needed -->

### Juju Events Changes

<!-- Any changes to the juju events being observed (newly added, significantly modified or deleted) -->

### Module Changes

<!-- Any high level changes to modules and why (Service, Observer, helper) -->

### Library Changes

<!-- Any changes to charm libraries -->

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

<!-- Explanation for any unchecked items above -->
charmhub doc will be updated as part of the documentation task

swetha1654 committed:
Port changes from 12-factor-changes

swetha1654 committed:
set status active

swetha1654 committed:
Merge branch 'main' into ISD-2485-actions

swetha1654 committed:
initial testing

swetha1654 committed:
delete and blacklist binaries

swetha1654 committed:
delete and blacklist binaries

swetha1654 committed:
delete and blacklist

swetha1654 committed:
add wrapper script

swetha1654 committed:
post testing

swetha1654 committed:
CI failure

swetha1654 committed:
move command execution code

swetha1654 committed:
generate src docs

swetha1654 committed:
shell check fail

swetha1654 committed:
integration test for delete

swetha1654 committed:
Post test changes

swetha1654 committed:
remove tracing library

yanksyoon commented on charmcraft.yaml:
```diff
@@ -35,27 +35,40 @@ requires:
+      - fingerprints
+      - ticket-id
+  rebuild-prefix-tree:
+    description: The prefix tree is used to manage the reconciliation process and 
+      must always be in sync with the database. If you suspect it has become corrupted or out 
+      of sync, or if the prefix tree needs to be removed and rebuilt after pg-restore, 
+      it can be done with this action. 
```
```suggestion
    description: |
      The prefix tree is used to manage the reconciliation process and 
      must always be in sync with the database. If you suspect it has become corrupted or out 
      of sync, or if the prefix tree needs to be removed and rebuilt after pg-restore, 
      it can be done with this action. 
```
May I suggest use of either `>` or `|` to keep it multiline friendly? :) 
https://yaml-multiline.info/

yanksyoon replied:
Same for other yaml declarations :)

swetha1654 replied:
addressed

yanksyoon commented on hockeypuck_rock/delete_keys:
Should the file be named `hockeypuck_rock/delete_keys.sh`?

yanksyoon replied:
Same for other scripts :D

swetha1654 replied:
addressed

yanksyoon commented on hockeypuck_rock/delete_keys:
```diff
@@ -0,0 +1,38 @@
+done
+
+$SQLCMD -c '
+  delete from subkeys k using deleted_keys b where k.rfingerprint = reverse(b.fingerprint);
+  alter table subkeys drop constraint subkeys_rfingerprint_fkey;
+  delete from keys k using deleted_keys b where k.rfingerprint = reverse(b.fingerprint);
+  alter table subkeys add foreign key (rfingerprint) references keys(rfingerprint);
```
Nit: Would it be helpful to capitalize SQL commands?
https://stackoverflow.com/questions/608196/why-should-i-capitalize-my-sql-keywords-is-there-a-good-reason
This is just to improve readability :)

seb4stien replied:
+1

swetha1654 replied:
addressed

yanksyoon commented on src/actions.py:
Out of curiosity, is there a reason for using shell scripts instead of using charm to handle the actions? I think the recommended way to handle things are to be as pythonic as possible :)

yanksyoon reviewed (commented):
Initial review, Looks good thanks!

seb4stien commented on charmcraft.yaml:
```diff
@@ -35,27 +35,40 @@ requires:
+      fingerprints:
+        type: string
+        description: Comma-separated list of full fingerprints to ignore (minus the leading 0x)
+      ticket-id:
```
I think we said in the project meeting we should avoid to be too canonical specific here. Wouldn't a "comment" field to the job? And then IS could use it to store the ticket id?

seb4stien commented on charmcraft.yaml:
```diff
@@ -35,27 +35,40 @@ requires:
     limit: 1
 
+actions:
+  blacklist-and-delete-key:
```
Shouldn't we turn it plural as you support multiple keys?

seb4stien commented on charmcraft.yaml:
```diff
@@ -35,27 +35,40 @@ requires:
 
+actions:
+  blacklist-and-delete-key:
+    description: Blacklist and delete a key fingerprint from the keyserver database.
```
To adapt if we turn plural.

seb4stien commented on hockeypuck_rock/hockeypuck_wrapper:
```diff
@@ -0,0 +1,20 @@
+
+SQLCMD="psql -t -A -d ${POSTGRESQL_DB_NAME} -h ${POSTGRESQL_DB_HOSTNAME} -U ${POSTGRESQL_DB_USERNAME}"
+
+$SQLCMD -c 'create table if not exists deleted_keys (fingerprint text primary key not null, comment text not null);'
```
You can also capitalize SQL commands here for readability.

seb4stien commented on pyproject.toml:
```diff
@@ -6,6 +6,7 @@ skips = ["*/*test.py", "*/test_*.py", "*tests/*.py"]
 # Testing tools configuration
 [tool.coverage.run]
 branch = true
+disable_warnings = ["no-data-collected"]
```
No sure why?

swetha1654 replied:
i deleted the unit test modules so when the unit test workflow runs it was throwing an error since there are no unit tests. I wasnt sure if i should delete the workflow or not so i added this. Should i delete the unit test github workflow? WDYT?

seb4stien replied:
To check with Greg on Wednesday.
It feels weird to not have any unit tests and to disable the associated warning or workflows.
I know we said we wouldn't have much to test here as we are using the 12-factors, but I wonder if we should not have at least one minimalistic unit test to stay aligned with our development practices.
But not strong opinion, I will follow Greg's point of view on this ;)

arturo-seijas replied:
Removing the workflow is definitely not the preferred option. The executed tox target the workflow runs also performs the static analysis checks and we'd also have to change the gates for merging PRs, deviating from the standard across all our charms, so I strongly recommend against doing that. IMO, the best option for these cases is always to tweak the project configuration: simpler, more obvious and more maintainable

swetha1654 replied:
As discussed in the standup, we will not be deleting the unit test workflow for now. Later if there are still no unit tests, it will be deleted. For now the disable warning will remain.

swetha1654 committed:
address review comments

swetha1654 committed:
rockcraft

swetha1654 committed:
rockcraft

swetha1654 committed:
post testing

swetha1654 committed:
add docstring

arturo-seijas commented on charmcraft.yaml:
```diff
@@ -35,27 +35,44 @@ requires:
     limit: 1
 
+actions:
+  blacklist-and-delete-keys:
```
Sounds a bit long. How about ` blacklist-keys`? I think it's safe to assume blacklisting will remove if needed

swetha1654 replied:
I'm not sure if changing it to blacklist-keys would be clear enough to the user unless they read the description. We can discuss this in the project meeting today

swetha1654 replied:
As discussed in the standup, changing this to block-key

arturo-seijas commented on charmcraft.yaml:
```diff
@@ -35,27 +35,44 @@ requires:
+  rebuild-prefix-tree:
+    description: |
+      The prefix tree is used to manage the reconciliation process and 
+      must always be in sync with the database. If you suspect it has become corrupted or out 
```
s/has become corrupted/ is corrupted. When writing technical documentation, you want to keep your grammar as simple as possible to that it is accessible to a wider audience :)

arturo-seijas commented on charmcraft.yaml:
```diff
@@ -35,27 +35,44 @@ requires:
         Comma-seperated list of external peer IPs or FQDNs without the ports.
-        Note that the http and reconcilliation ports are fixed and cannot be
-        edited.
+        Note that the http and reconcilliation ports are fixed and cannot be edited.
```
s/http/HTTP

arturo-seijas commented on charmcraft.yaml:
```diff
@@ -35,27 +35,44 @@ requires:
-        sign administration requests for this server. It is strongly
-        recommended that the server contact advertised on the stats page
-        is not one of the admin keys.
+        sign administration requests for this server. It is strongly recommended that the server 
```
s/sign/Sign

arturo-seijas replied:
Does "It is strongly recommended that the server contact advertised on the stats page is not one of the admin keys." apply here?

swetha1654 replied:
i dont think sign should be capitalized, its not the beginning of the sentence

swetha1654 replied:
removed the next sentence, not required

arturo-seijas replied:
Indeed. My bad

arturo-seijas commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,97 @@
+        conn = psycopg2.connect(dsn)
+        conn.autocommit = True
+        return conn
+    except Exception as e:
```
You should avoid catching generic exceptions, as you might catch something unexpected

arturo-seijas commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,97 @@
+        conn.autocommit = True
+        return conn
+    except Exception as e:
+        print(f"Failed to connect to database: {e}")
```
Shouldn't you log instead of printing?

seb4stien commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,97 @@
+        cursor: the database cursor.
+        fingerprints: list of fingerprints to delete.
+        comment: the comment associated with the deletion."""
+    logging.info(f"Deleting fingerprints: {', '.join(fingerprints)}")
```
It's better to avoid f-strings for logs (for performance and security reasons: https://google.github.io/styleguide/pyguide.html#3101-logging)

arturo-seijas replied:
The `%` operator is the recommended way to interpolate strings when logging https://stackoverflow.com/questions/54367975/python-3-7-logging-f-strings-vs

arturo-seijas commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,97 @@
+        comment: the comment associated with the deletion."""
+    logging.info(f"Deleting fingerprints: {', '.join(fingerprints)}")
+    try:
+        for fp in fingerprints:
```
s/fp/fingerprint

arturo-seijas commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,97 @@
+            REFERENCES keys(rfingerprint);
+            """
+        )
+        print("Deletion process completed successfully.")
```
Same here

arturo-seijas commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,97 @@
+logger = logging.getLogger(__name__)
+
+
+def get_db_connection():
```
You are missing the return type hints for all the method in this class

swetha1654 replied:
added them now

seb4stien commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,97 @@
+            WHERE subkeys.rfingerprint = REVERSE(deleted_keys.fingerprint);
+        """
+        )
+        cursor.execute("ALTER TABLE subkeys DROP CONSTRAINT IF EXISTS subkeys_rfingerprint_fkey;")
```
I guess it comes from an existing script?
If feel a bit weird to drop a consistency contraint. Don't we have a way to delete the key and the subkeys in a single transaction to avoid droping the foreign key relationship?

seb4stien replied:
I saw that you re-add the constraint below. I'm still wondering if this couldn't be avoided ;)

swetha1654 replied:
you're right, i took these SQL commands from the delete script provided by the hockeypuck developers. I'm guessing the reason they are dropping and re-adding the constraint is to prevent any deletion failures due to constraint issues. Even though i haven't personally seen any consequences to removing those constraint SQL commands, i would prefer running them anyway since they are present in the delete script provided by the hockeypuck developers.

seb4stien commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,97 @@
+    )
+    parser.add_argument("--comment", required=True, help="Comment associated with the deletion")
+    args = parser.parse_args()
+    fingerprints = args.fingerprints.split(",")
```
Shouldn't we have some format validation on the fingerprints?

swetha1654 replied:
that's a good idea, i have added input validation to see if we are being passed valid fingerprints

arturo-seijas commented on src/actions.py:
```diff
@@ -0,0 +1,91 @@
+
+logger = logging.getLogger(__name__)
+
+HOCKEYPUCK_CONTAINER_NAME = "app"
```
Isn't this constant defined on the PaaS dependency?

swetha1654 replied:
you're right, removed this

arturo-seijas commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+
+# Copyright 2025 Canonical Ltd.
```
Is this file being linted? Because it doesn't look like it is

swetha1654 replied:
it got missed since it was in hockeypuck_rock folder. Added it to the list of files to be linted in tox.ini

swetha1654 commented on src/actions.py:
```diff
@@ -0,0 +1,91 @@
+
+"""Hockeypuck charm actions."""
+
+# pylint: disable=protected-access
```
i needed this since i am accessing a private method from the paas_app_charmer (_gen_environment), but since it is just one instance i have moved this pylint disable to just that line instead of disabling it for the entire class

swetha1654 committed:
input validation

swetha1654 committed:
change action name

swetha1654 committed:
change action name

arturo-seijas commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,120 @@
+    try:
+        delete_fingerprints(cursor, fingerprints, comment)
+    finally:
+        cursor.close()
```
You could use `with` instead of explicitly closing the resources

arturo-seijas commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,120 @@
+logger = logging.getLogger(__name__)
+
+
+def get_db_connection() -> psycopg2.extensions.connection | None:
```
I'd rather re-raise the exception rather than returning a None value since it would be an abnormal flow

swetha1654 replied:
addressed in the next [PR](https://github.com/canonical/hockeypuck-k8s-operator/pull/12)

arturo-seijas commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,120 @@
+        return None
+
+
+def delete_fingerprints(
```
Same here

swetha1654 replied:
addressed in the next [PR](https://github.com/canonical/hockeypuck-k8s-operator/pull/12)

arturo-seijas commented on hockeypuck_rock/delete_keys.py:
```diff
@@ -0,0 +1,120 @@
+    try:
+        delete_fingerprints(cursor, fingerprints, comment)
+    finally:
+        cursor.close()
```
If you raised an exception, you could just let it exit the execution on its own

swetha1654 replied:
addressed in the next [PR](https://github.com/canonical/hockeypuck-k8s-operator/pull/12)

arturo-seijas commented on src/actions.py:
```diff
@@ -62,29 +59,23 @@ def _rebuild_prefix_tree_action(self, event: ops.ActionEvent) -> None:
             process = hockeypuck_container.exec(
                 command,
-                environment=environment,
+                environment=self.charm._gen_environment(),  # pylint: disable=protected-access
```
Same here

swetha1654 replied:
changed this in the next PR to access a public method

swetha1654 commented on src/actions.py:
```diff
@@ -62,29 +59,23 @@ def _rebuild_prefix_tree_action(self, event: ops.ActionEvent) -> None:
         if not self.charm.is_ready():
             event.fail("Service not yet ready.")
-        hockeypuck_container = self.charm.unit.containers[HOCKEYPUCK_CONTAINER_NAME]
+        hockeypuck_container = self.charm._container  # pylint: disable=protected-access
```
changed this in the next PR to access a public method

swetha1654 commented:
Closing this PR since I had opened another [PR](https://github.com/canonical/hockeypuck-k8s-operator/pull/12) simultaneously and both PRs have diverged significantly. I have ported the changes in this PR addressed the review comments here in the next PR.
